File: scripts/gen_update_manifests.py
from __future__ import annotations

import json
import os
import re
from datetime import datetime, timezone
from typing import Any

import mkdocs_gen_files
import requests

OWNER = "strantalis"
REPO = "workset"
API_URL = f"https://api.github.com/repos/{OWNER}/{REPO}/releases"
USER_AGENT = "workset-docs-updater-manifests"
CHANNELS = ("stable", "alpha")
SHA256_RE = re.compile(r"\b([a-fA-F0-9]{64})\b")
TEAM_ID_RE = re.compile(r"\b([A-Z0-9]{10})\b")


def _now_rfc3339() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def _base_headers() -> dict[str, str]:
    headers = {
        "Accept": "application/vnd.github+json",
        "User-Agent": USER_AGENT,
    }
    token = os.getenv("GITHUB_TOKEN") or os.getenv("GH_TOKEN")
    if token:
        headers["Authorization"] = f"Bearer {token}"
    return headers


def _fetch_releases() -> tuple[list[dict[str, Any]], str | None]:
    try:
        response = requests.get(
            API_URL,
            headers=_base_headers(),
            timeout=20,
            params={"per_page": 100},
        )
        response.raise_for_status()
        data = response.json()
    except Exception as exc:  # pragma: no cover - best-effort docs generation
        return [], str(exc)
    return [r for r in data if not r.get("draft")], None


def _select_release(releases: list[dict[str, Any]], channel: str) -> dict[str, Any] | None:
    prerelease = channel == "alpha"
    for release in releases:
        if bool(release.get("prerelease")) == prerelease:
            return release
    return None


def _fetch_text(url: str) -> tuple[str, str | None]:
    try:
        response = requests.get(url, headers=_base_headers(), timeout=30)
        response.raise_for_status()
        return response.text, None
    except Exception as exc:  # pragma: no cover - best-effort docs generation
        return "", str(exc)


def _default_latest(version: str, notes_url: str, asset_url: str, generated_at: str) -> dict[str, Any]:
    return {
        "version": version,
        "pubDate": generated_at,
        "notesUrl": notes_url,
        "minimumVersion": "",
        "asset": {
            "name": "disabled",
            "url": asset_url,
            "sha256": "disabled",
        },
        "signing": {
            "teamId": "DISABLED",
        },
    }


def _release_tag(release: dict[str, Any] | None) -> str:
    if not release:
        return "v0.0.0"
    tag = str(release.get("tag_name") or "").strip()
    if not tag:
        return "v0.0.0"
    if not tag.startswith("v"):
        return f"v{tag}"
    return tag


def _build_manifest(
    channel: str,
    release: dict[str, Any] | None,
    error: str | None,
) -> dict[str, Any]:
    generated_at = _now_rfc3339()
    notes_url = str((release or {}).get("html_url") or "https://workset.dev/releases/")
    default = {
        "schemaVersion": 1,
        "generatedAt": generated_at,
        "channel": channel,
        "disabled": True,
        "message": "",
        "latest": _default_latest(_release_tag(release), notes_url, notes_url, generated_at),
    }

    if error:
        default["message"] = f"Update manifest generation failed: {error}"
        return default

    if not release:
        default["message"] = f"No {channel} release available yet."
        return default

    assets = release.get("assets") or []
    assets_by_name = {
        str(asset.get("name") or ""): asset for asset in assets if asset.get("name")
    }

    tag = _release_tag(release)
    default["latest"]["version"] = tag
    default["latest"]["pubDate"] = str(release.get("published_at") or generated_at)
    default["latest"]["notesUrl"] = notes_url

    expected_zip_name = f"workset-{tag}-macos-update.zip"
    update_asset = assets_by_name.get(expected_zip_name)
    if not update_asset:
        default["message"] = (
            f"Latest {channel} release {tag} does not include a macOS auto-update package yet."
        )
        return default

    update_asset_url = str(update_asset.get("browser_download_url") or "").strip()
    if not update_asset_url:
        default["message"] = f"Latest {channel} release {tag} has an invalid update asset URL."
        return default

    sha_asset = assets_by_name.get(f"{expected_zip_name}.sha256")
    team_asset = assets_by_name.get(f"{expected_zip_name}.teamid")
    if not sha_asset or not team_asset:
        default["latest"]["asset"]["name"] = expected_zip_name
        default["latest"]["asset"]["url"] = update_asset_url
        default["message"] = (
            f"Latest {channel} release {tag} is missing updater metadata assets (.sha256/.teamid)."
        )
        return default

    sha_text, sha_error = _fetch_text(str(sha_asset.get("browser_download_url") or ""))
    team_text, team_error = _fetch_text(str(team_asset.get("browser_download_url") or ""))
    if sha_error or team_error:
        default["latest"]["asset"]["name"] = expected_zip_name
        default["latest"]["asset"]["url"] = update_asset_url
        default["message"] = (
            f"Failed to load updater metadata for {tag}: "
            f"sha256={sha_error or 'ok'}, teamid={team_error or 'ok'}."
        )
        return default

    sha_match = SHA256_RE.search(sha_text)
    team_match = TEAM_ID_RE.search(team_text)
    if not sha_match or not team_match:
        default["latest"]["asset"]["name"] = expected_zip_name
        default["latest"]["asset"]["url"] = update_asset_url
        default["message"] = f"Updater metadata for {tag} is invalid."
        return default

    default["disabled"] = False
    default["message"] = ""
    default["latest"]["asset"]["name"] = expected_zip_name
    default["latest"]["asset"]["url"] = update_asset_url
    default["latest"]["asset"]["sha256"] = sha_match.group(1).lower()
    default["latest"]["signing"]["teamId"] = team_match.group(1)
    return default


releases, release_error = _fetch_releases()

for channel in CHANNELS:
    release = _select_release(releases, channel)
    manifest = _build_manifest(channel, release, release_error)
    with mkdocs_gen_files.open(f"updates/{channel}.json", "w") as fd:
        json.dump(manifest, fd, indent=2)
        fd.write("\n")

File: docs-site/scripts/gen-update-manifests.mjs
/**
 * Generate update manifest JSON files for the in-app updater.
 * Outputs updates/stable.json and updates/alpha.json into public/.
 *
 * Ported from scripts/gen_update_manifests.py (MkDocs era).
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const OWNER = 'strantalis';
const REPO = 'workset';
const API_URL = `https://api.github.com/repos/${OWNER}/${REPO}/releases`;
const CHANNELS = ['stable', 'alpha'];
const SHA256_RE = /\b([a-fA-F0-9]{64})\b/;
const TEAM_ID_RE = /\b([A-Z0-9]{10})\b/;

function nowRFC3339() {
  return new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
}

function headers() {
  const h = {
    Accept: 'application/vnd.github+json',
    'User-Agent': 'workset-docs-updater-manifests',
  };
  const token = process.env.GITHUB_TOKEN || process.env.GH_TOKEN;
  if (token) h.Authorization = `Bearer ${token}`;
  return h;
}

async function fetchReleases() {
  try {
    const res = await fetch(`${API_URL}?per_page=100`, { headers: headers() });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const data = await res.json();
    return [data.filter((r) => !r.draft), null];
  } catch (e) {
    return [[], e.message];
  }
}

function selectRelease(releases, channel) {
  const prerelease = channel === 'alpha';
  let fallback = null;
  for (const release of releases) {
    if (Boolean(release.prerelease) !== prerelease) continue;
    if (!fallback) fallback = release;
    const tag = releaseTag(release);
    const names = new Set((release.assets || []).map((a) => a.name).filter(Boolean));
    const zip = `workset-${tag}-macos-update.zip`;
    if (names.has(zip) && names.has(`${zip}.sha256`) && names.has(`${zip}.teamid`)) {
      return release;
    }
  }
  return fallback;
}

async function fetchText(url) {
  try {
    const res = await fetch(url, { headers: headers() });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    return [await res.text(), null];
  } catch (e) {
    return ['', e.message];
  }
}

function releaseTag(release) {
  if (!release) return 'v0.0.0';
  let tag = (release.tag_name || '').trim();
  if (!tag) return 'v0.0.0';
  if (!tag.startsWith('v')) tag = `v${tag}`;
  return tag;
}

function defaultLatest(version, notesUrl, assetUrl, generatedAt) {
  return {
    version,
    pubDate: generatedAt,
    notesUrl,
    minimumVersion: '',
    asset: { name: 'disabled', url: assetUrl, sha256: 'disabled' },
    signing: { teamId: 'DISABLED' },
  };
}

async function buildManifest(channel, release, error) {
  const generatedAt = nowRFC3339();
  const notesUrl = release?.html_url || 'https://workset.dev/releases/';
  const tag = releaseTag(release);
  const manifest = {
    schemaVersion: 1,
    generatedAt,
    channel,
    disabled: true,
    message: '',
    latest: defaultLatest(tag, notesUrl, notesUrl, generatedAt),
  };

  if (error) {
    manifest.message = `Update manifest generation failed: ${error}`;
    return manifest;
  }
  if (!release) {
    manifest.message = `No ${channel} release available yet.`;
    return manifest;
  }

  manifest.latest.version = tag;
  manifest.latest.pubDate = release.published_at || generatedAt;
  manifest.latest.notesUrl = notesUrl;

  const assetsByName = Object.fromEntries(
    (release.assets || []).filter((a) => a.name).map((a) => [a.name, a])
  );

  const zipName = `workset-${tag}-macos-update.zip`;
  const updateAsset = assetsByName[zipName];
  if (!updateAsset) {
    manifest.message = `Latest ${channel} release ${tag} does not include a macOS auto-update package yet.`;
    return manifest;
  }

  const updateURL = (updateAsset.browser_download_url || '').trim();
  if (!updateURL) {
    manifest.message = `Latest ${channel} release ${tag} has an invalid update asset URL.`;
    return manifest;
  }

  const shaAsset = assetsByName[`${zipName}.sha256`];
  const teamAsset = assetsByName[`${zipName}.teamid`];
  if (!shaAsset || !teamAsset) {
    manifest.latest.asset.name = zipName;
    manifest.latest.asset.url = updateURL;
    manifest.message = `Latest ${channel} release ${tag} is missing updater metadata assets (.sha256/.teamid).`;
    return manifest;
  }

  const [shaText, shaErr] = await fetchText(shaAsset.browser_download_url);
  const [teamText, teamErr] = await fetchText(teamAsset.browser_download_url);
  if (shaErr || teamErr) {
    manifest.latest.asset.name = zipName;
    manifest.latest.asset.url = updateURL;
    manifest.message = `Failed to load updater metadata for ${tag}: sha256=${shaErr || 'ok'}, teamid=${teamErr || 'ok'}.`;
    return manifest;
  }

  const shaMatch = SHA256_RE.exec(shaText);
  const teamMatch = TEAM_ID_RE.exec(teamText);
  if (!shaMatch || !teamMatch) {
    manifest.latest.asset.name = zipName;
    manifest.latest.asset.url = updateURL;
    manifest.message = `Updater metadata for ${tag} is invalid.`;
    return manifest;
  }

  manifest.disabled = false;
  manifest.message = '';
  manifest.latest.asset.name = zipName;
  manifest.latest.asset.url = updateURL;
  manifest.latest.asset.sha256 = shaMatch[1].toLowerCase();
  manifest.latest.signing.teamId = teamMatch[1];
  return manifest;
}

async function main() {
  const outDir = path.resolve(__dirname, '..', 'public', 'updates');
  fs.mkdirSync(outDir, { recursive: true });

  const [releases, error] = await fetchReleases();
  for (const channel of CHANNELS) {
    const release = selectRelease(releases, channel);
    const manifest = await buildManifest(channel, release, error);
    const outPath = path.join(outDir, `${channel}.json`);
    fs.writeFileSync(outPath, JSON.stringify(manifest, null, 2) + '\n');
    console.log(`wrote ${outPath} (${manifest.disabled ? 'disabled' : manifest.latest.version})`);
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
